quinn-proto/src/frame.rs

```diff
@@ -704,7 +703,7 @@ fn scan_ack_blocks(packet: &[u8], largest: u64, n: usize) -> Option<usize> {
         let block = buf.get_var().ok()?;
         smallest = smallest.checked_sub(block)?;
     }
-    Some(buf.position() as usize)
+    Some((buf.position() - start) as usize)
```

Comment:
Should definitely be split out, have done so now.

It's not quite as straightforward as what you're suggesting because of the `checked_sub()?` calls, which need to propagate `None` in order to keep it easy to read. Could return `Option<()>` or create a new error wrapper that has `UnexpectedEnd` and `InvalidSub` or whatever, but that doesn't seem so nice. So I've otherwise kept it as is.

Reply:
I was thinking we'd write out `.ok_or(UnexpectedEnd)` for those cases, accepting a slightly more verbose (but readable) implementation for a simpler interface.

Reply:
What I dislike about that is that `UnexpectedEnd` is incorrect for the errors arising from `checked_sub()`...

Reply:
Oh, right; this could be `IterErr` as used by the caller, which already has the necessary cases.

Reply:
Nice one, fixed that up.